packages/@lwc/template-compiler/src/codegen/helpers.ts

```diff
@@ -88,30 +84,6 @@ export function hasIdAttribute(node: Node): boolean {
     return false;
 }
 
-export function memorizeHandler(
-    codeGen: CodeGen,
-    componentHandler: t.Expression,
-    handler: t.Expression
-): t.Expression {
-    // #439 - The handler can only be memorized if it is bound to component instance
-    const id = getMemberExpressionRoot(componentHandler as t.MemberExpression);
-    const shouldMemorizeHandler = !codeGen.isLocalIdentifier(id);
-
-    // Apply memorization if the handler is memorizable.
-    //   $cmp.handlePress -> _m1 || ($ctx._m1 = b($cmp.handlePress))
-    if (shouldMemorizeHandler) {
```

Comment:
The original code has a `shouldMemorizeHandler` check here. The goal is to prevent memoization in cases where the listener is referenced in a scope outside of the component scope, e.g. inside of a `for:each:

https://github.com/salesforce/lwc/blob/39af4df1fd18d2d4835fcc532795b9b65e1ba23b/packages/%40lwc/template-compiler/src/__tests__/fixtures/regression/handler-memoization/actual.html#L4-L9

In this case, we can't memoize the `onclick` (or `on`) because `task` may change between renders.

This behavior is a bit subtle, so I added a new test in a separate PR: #4480

Reply:
I think for the purposes of this PR we can just check if every listener in the `listenerObjectAST` is memoizable. If a single one is not memoizable, then we can bail out of the optimization for the entire `on` object.